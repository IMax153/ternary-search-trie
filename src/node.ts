/**
 * Represents a single node in the ternary search trie.
 */
export class Node<Value> {
  /**
   * The key assigned to the node.
   */
  public readonly key: string;

  /**
   * The value assigned to the node.
   */
  public value: Value | null;

  /**
   * The left child of the node. Has a key value
   * that is less than the key value of its parent.
   */
  public left: Node<Value> | null;

  /**
   * The middle child of the node. Has a key value
   * that is equivalent to the key value of its parent.
   */
  public middle: Node<Value> | null;

  /**
   * The right child of the node. Has a key value
   * that is greater than the key value of its parent.
   */
  public right: Node<Value> | null;

  /**
   * The parent of the node.
   */
  public parent: Node<Value> | null;

  public constructor(key: string, value: Value | null = null) {
    this.key = key;
    this.value = value;
    this.parent = null;
    this.left = null;
    this.middle = null;
    this.right = null;
  }
}
