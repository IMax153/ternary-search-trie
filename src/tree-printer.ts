/**
 * Optional parameters for printing the tree.
 * @typedef {Object} TreePrinterOptions
 * @param {boolean} [showFunctions=false] - Determines whether or not functions should be displayed.
 * @param {boolean} [showValues=false] - Determines whether or not values should be displayed.
 * @param {string[]} [skipKeys] - Keys that should not be displayed.
 */
export interface TreePrinterOptions {
  showFunctions?: boolean;
  showValues?: boolean;
  skipKeys?: string[];
}

/**
 * A class that handles printing of trees in an easy-to-read format.
 */
export class TreePrinter<T extends Record<string, any>> {
  /**
   *
   * @param {T} tree - The tree to print.
   * @param {TreePrinterOptions} options - Optional parameters for printing the tree.
   */
  public static create<T>(tree: T, options: TreePrinterOptions): TreePrinter<T> {
    return new TreePrinter(tree, options);
  }

  /**
   * The tree that will be printed.
   */
  private obj: T;

  /**
   * Determines whether or not functions should be displayed.
   */
  private showFunctions: boolean;

  /**
   * Determines whether or not values should be displayed.
   */
  private showValues: boolean;

  /**
   * Keys that should not be displayed.
   */
  private skipKeys: string[];

  private constructor(
    obj: T,
    { showFunctions = false, showValues = false, skipKeys = [] }: TreePrinterOptions,
  ) {
    this.obj = obj;
    this.showFunctions = showFunctions;
    this.showValues = showValues;
    this.skipKeys = skipKeys;
  }

  /**
   * Gets a string representation of the entire object as a tree line-by-line and passes each line to the specified callback function.
   * @param {(line: string) => void} lineCallback - The callback that will be fired with each line.
   */
  public asLines(lineCallback: (line: string) => void): void {
    this.growBranch(
      '.',
      this.obj,
      false,
      [],
      this.showValues,
      this.showFunctions,
      this.skipKeys,
      lineCallback,
    );
  }

  /**
   * Gets a string representation of the entire object as a tree with line breaks.
   */
  public asTree(): string {
    let tree = '';

    this.growBranch(
      '.',
      this.obj,
      false,
      [],
      this.showValues,
      this.showFunctions,
      this.skipKeys,
      line => {
        tree += `${line}\n`;
      },
    );

    return tree;
  }

  /**
   * Creates a prefix for the tree branch.
   * @param hasKey - Determines if there is a key.
   * @param isTerminal - Determines if this is a terminal key.
   */
  private makePrefix(hasKey: boolean, isTerminal: boolean) {
    const str = isTerminal ? '└' : '├';
    return hasKey ? str.concat('─ ') : str.concat('──┐');
  }

  /**
   * Filters the keys of the specified object.
   * @param obj - The object whose keys to filter.
   * @param hideFunctions - Determines if functions should be displayed.
   * @param skipKeys - Determins which keys to skip.
   */
  private filterKeys(obj: T, hideFunctions: boolean, skipKeys: string[]): string[] {
    const keys: string[] = [];

    if (!obj) return keys;

    Object.keys(obj).forEach((branch: string) => {
      let isFunction = false;

      // Only display values that are not in the Object prototype chain
      const hasProp = Object.prototype.hasOwnProperty.call(obj, branch);

      if (hasProp) {
        isFunction = typeof obj[branch] === 'function';
      }

      // Determine if the current item should be shown
      const shouldDisplay = !skipKeys.some(key => key === branch) || !(hideFunctions && isFunction);

      if (hasProp || shouldDisplay) {
        keys.push(branch);
      }
    });

    return keys;
  }

  /**
   * Recursively grows a branch in the tree until all subtrees are resolved.
   * @param {string} key - The key of the branch.
   * @param {Object} root - The root of the branch.
   * @param {boolean} isTerminal - Determines if the branch is terminal.
   * @param {(T,boolean)[]} lastStates - The previous branches states.
   * @param {boolean} showValues - Determines if values should be displayed.
   * @param {boolean} hideFunctions - Determines if values should be displayed.
   * @param {string[]} skipKeys - Which keys should be hidden if any.
   * @param {(key: string) => void} callback - Callback with the current key value.
   */
  private growBranch(
    key: keyof T,
    root: T,
    isTerminal: boolean,
    lastStates: [T, boolean][],
    showValues: boolean,
    hideFunctions: boolean,
    skipKeys: string[],
    callback: (line: string) => void,
  ) {
    let line = '';
    let index = 0;
    let lastKey = false;
    let circular = false;
    const lastStatesCopy = lastStates.slice(0);

    if (lastStatesCopy.push([root, isTerminal]) && lastStates.length > 0) {
      // Based on the `isTerminal` state of whatever we're nested within,
      // we need to append either blankness or a branch to our line
      lastStates.forEach((lastState, idx) => {
        if (idx > 0) {
          line += `${lastState[1] ? ' ' : '│'}  `;
        }

        if (!circular && lastState[0] === root) {
          circular = true;
        }
      });

      // The generated prefix varies based on whether the key contains something to show and
      // whether we're dealing with the last element in this collection
      line += this.makePrefix(!!key, isTerminal) + key;

      if (showValues && (typeof root !== 'object' || root instanceof Date)) {
        line += `: ${root}`;
      }

      if (circular) {
        line += ' (circular ref.)';
      }

      callback(line);
    }

    // Can we descend into the next item?
    if (!circular && typeof root === 'object') {
      const keys = this.filterKeys(root, hideFunctions, skipKeys);

      keys.forEach(branch => {
        // The last key is always printed with a different prefix, so we'll need to know if we have it
        lastKey = ++index === keys.length;

        this.growBranch(
          branch,
          root[branch],
          lastKey,
          lastStatesCopy,
          showValues,
          hideFunctions,
          skipKeys,
          callback,
        );
      });
    }
  }
}
