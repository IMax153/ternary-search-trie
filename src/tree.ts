import { PrintTree, PrintTreeOptions } from './print-tree';

/**
 * Represents a single node in the ternary search trie.
 */
class TrieNode<Value> {
  /**
   * The key of the node.
   */
  public readonly key: string;

  /**
   * The value of the node.
   */
  public value: Value | null;

  /**
   * The parent of the node.
   */
  public parent: TrieNode<Value> | null;

  /**
   * The left subtree. Has a lower key value.
   */
  public left: TrieNode<Value> | null;

  /**
   * The left subtree. Has an equivalent key value
   */
  public middle: TrieNode<Value> | null;

  /**
   * The left subtree. Has a higher key value.
   */
  public right: TrieNode<Value> | null;

  public constructor(key: string) {
    this.key = key;
    this.value = null;
    this.parent = null;
    this.left = null;
    this.middle = null;
    this.right = null;
  }

  /**
   * Determines if the node has a left child.
   */
  public hasLeftChild(): boolean {
    return !!this.left;
  }

  /**
   * Determines if the node has a middle child.
   */
  public hasMiddleChild(): boolean {
    return !!this.middle;
  }

  /**
   * Determines if the node has a right child.
   */
  public hasRightChild(): boolean {
    return !!this.right;
  }

  /**
   * Determines if the node is a parent for any subtrees.
   */
  public hasChildren(): boolean {
    return this.hasLeftChild() || this.hasMiddleChild() || this.hasRightChild();
  }
}

/**
 * Represents a ternary search trie.
 * https://en.wikipedia.org/wiki/Ternary_search_tree
 */
export class Trie<Value> {
  /**
   * The root node of the tree.
   */
  private root: TrieNode<Value> | null;

  /**
   * The number of nodes within the tree.
   */
  private nodeCount: number;

  /**
   * Gets the size of the tree in terms of the number of nodes present.
   */
  public get size(): number {
    return this.nodeCount;
  }

  public constructor() {
    this.root = null;
    this.nodeCount = 0;
  }

  /**
   * Gets all keys present in the tree.
   */
  public keys(): string[] {
    const found: string[] = [];

    this.traversal(this.root, node => {
      found.push(node.key);
    });

    return found;
  }

  /**
   * Creates a new node in the tree with the specified key and value.
   * @param {string} key - The key of the node.
   * @param {Value} value - The value of the node.
   */
  public set(key: string, value: Value): Trie<Value> {
    this.validateInput(key);

    if (!this.root) {
      this.root = new TrieNode<Value>([...key][0]);
    }

    this.insert(key, value, this.root);

    return this;
  }

  /**
   * Recursively searches the tree for the node with the specified key and returns its value, if possible.
   * @param {string} key - The key of the node.
   * @param {boolean} splay - // TODO: FIGURE THIS SHIT OUT
   */
  public get(key: string, splay?: boolean): Value | null {
    this.validateInput(key);
    const node = splay ? new TrieNode<Value>(key) : this.find(key, this.root);
    return node ? node.value : null;
  }

  /**
   * Deletes a node from the tree.
   * @param {string} key - The key of the node to delete.
   */
  public del(key: string): Trie<Value> {
    this.validateInput(key);

    const node = this.find(key, this.root);

    if (node) {
      node.value = null;
      this.delete(node);
      this.nodeCount--;
    }

    return this;
  }

  /**
   * Performs a depth-first traversal of the tree starting from the root node.
   * @param {function(TrieNode<Value>): void} callback - The callback to execute at each visited node.
   */
  public traverse(callback: (node: TrieNode<Value>) => void) {
    this.traversal(this.root, callback);
  }

  /**
   * Searches the tree using the specified prefix.
   * @param {string} prefix - The prefix to search by.
   * @param {(TrieNode<Value>): void} callback - The callback to execute at each visited node.
   */
  public searchByPrefix(
    prefix: string,
    callback: (prefix: string, root: TrieNode<Value>) => void,
  ): Trie<Value> {
    this.validateInput(prefix);

    const root = this.find(prefix, this.root);

    if (!root) return this;

    // If the node we are searching from has a value (i.e. there is a one letter word
    // beginning with the specified prefix), execute the callback
    if (root.value) callback(prefix, root);

    // All words should start with the root node found using the prefix
    let word = [root.key];

    // Traversal starts from the middle child - we only care about words that start with the specified prefix
    this.traversal(root.middle, node => {
      // Add each node's key found during the traversal to the word
      word.push(node.key);

      // If a value is found at a given node, execute the callback
      if (node.value) {
        callback(word.join(''), node);

        // If the found node has any children, the current letter must be popped off the stack
        if (node.hasChildren()) {
          word.pop();
        } else {
          // Otherwise we reset the word to the root prefix and allow traversal to continue
          word = [root.key];
        }
      }
    });

    return this;
  }

  /**
   * Creates a string representation of the tree.
   */
  public toString(options: PrintTreeOptions = {}): string {
    const treeify = PrintTree.create(this.root, options);
    return treeify.asTree();
  }

  /**
   * Recursively searches the tree for the node with the specified key and returns its value, if possible.
   * @param {string} word - The key of the node to find.
   * @param {TrieNode<Value>} root - The node to start the search from.
   */
  private find(word: string, node: TrieNode<Value> | null): TrieNode<Value> | null {
    const key = [...word][0];

    if (!node) return null;

    // Recurse down the left subtree
    if (key < node.key) {
      return this.find(word, node.left);
    }

    // Recurse down the middle subtree.
    if (key === node.key) {
      // If there are still characters in the key, move the search cursor forward by one.
      if (word.length > 1) {
        return this.find(word.slice(1), node.middle);
      }

      return node;
    }

    // Recurse down the right subtree
    return this.find(word, node.right);
  }

  /**
   * Recursively inserts a word into the tree. Each character in the word becomes a node in the tree.
   * The the specified value is set to the terminal node of the word.
   * @param {string} word - The word to insert.
   * @param {TrieNode<Value>} node - The node to insert.
   * @param {Value} value - The value for the node.
   */
  private insert(word: string, value: Value, node: TrieNode<Value> | null): TrieNode<Value> | null {
    const key = [...word][0];

    if (!key) return null;

    if (!node) {
      node = new TrieNode(key);
    }

    if (!this.root) {
      this.root = node;
      this.nodeCount++;
    }

    if (key < node.key) {
      node.left = this.insert(word, value, node.left);

      if (node.left) node.left.parent = node;
    } else if (key === node.key) {
      if (word.length > 1) {
        node.middle = this.insert(word.slice(1), value, node.middle);

        if (node.middle) node.middle.parent = node;
      } else {
        node.value = value;
        this.nodeCount++;
      }
    } else {
      node.right = this.insert(word, value, node.right);

      if (node.right) node.right.parent = node;
    }

    return node;
  }

  /**
   * Recursively deletes a node from the tree if the node has no left or right sub-tree.
   * @param {TrieNode<Value>} node - The node to delete.
   */
  private delete(node: TrieNode<Value>) {
    if (!node) return;

    const { parent, left, middle, right } = node;

    if (middle) return;

    const hasUnequalChildren = !!(left || right);

    if (!hasUnequalChildren) {
      if (parent) {
        this.erase(parent, node);
        this.delete(parent);
      } else {
        const child = left || right;
        this.root = child;
        if (this.root) this.root.parent = null;
      }
    }
  }

  /**
   * Validates input to the tree.
   * @param {string} input - The input value.
   */
  private validateInput(input: string) {
    if (typeof input !== 'string') {
      throw new TypeError(
        `[Trie] Only string values can be used as keys. Received type ${typeof input}.`,
      );
    }

    if (input.length < 1) {
      throw new Error(`[Trie] Key values must have length greater than or equal to 1.`);
    }
  }

  /**
   * Erases a node from the tree.
   * @param {TrieNode<Value>} node - The node to erase.
   * @param {TrieNode<Value>} parent - The parent of the node to erase.
   */
  private erase(node: TrieNode<Value>, parent: TrieNode<Value>) {
    node.parent = null;

    switch (node) {
      case parent.left:
        parent.left = null;
        break;

      case parent.middle:
        parent.middle = null;
        break;

      case parent.right:
        parent.right = null;
        break;

      default:
        break;
    }
  }

  /**
   * Performs a depth-first traversal of the tree.
   * @param {TrieNode<Value>} root - The node to start the traversal from.
   * @param {function(TrieNode<Value>): void} callback - The callback to execute at each visited node.
   */
  private traversal(root: TrieNode<Value> | null, callback: (node: TrieNode<Value>) => void): void {
    if (!root) return;

    callback(root);

    if (root.left) {
      this.traversal(root.left, callback);
    }

    if (root.middle) {
      this.traversal(root.middle, callback);
    }

    if (root.right) {
      this.traversal(root.right, callback);
    }
  }
}
