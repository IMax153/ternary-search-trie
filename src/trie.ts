import { inspect } from 'util';

import { Node } from './node';

/**
 * Represents a ternary search trie.
 * https://en.wikipedia.org/wiki/Ternary_search_tree
 */
export class Trie<Value> {
  /**
   * The root node of the tree.
   */
  private root: Node<Value> | null;

  /**
   * The number of nodes within the tree.
   */
  private count: number;

  public constructor() {
    this.root = null;
    this.count = 0;
  }

  /**
   * Returns true if the tree contains any nodes, otherwise false.
   */
  public get isEmpty(): boolean {
    return this.count === 0;
  }

  /**
   * Returns the number of words that have been inserted into the tree.
   */
  public get size(): number {
    return this.count;
  }

  /**
   * Adds the specified key/value pair to the tree.
   * @param key The key to add.
   * @param value The value of the key.
   */
  public set(key: string, value: Value): this {
    this.validateKey(key);

    this.insert(this.root, key, value);

    return this;
  }

  /**
   * Returns the value of the node with the specified key.
   * @param key The key of the node.
   */
  public get(key: string): Value | null {
    this.validateKey(key);

    const node = this.find(key, this.root);

    return node?.value ?? null;
  }

  /**
   * Deletes the node with the specified key.
   * @param key The key of the node to delete.
   */
  public del(key: string): this {
    const node = this.find(key, this.root);

    if (node) {
      node.value = null;
      this.delete(node);
    }

    return this;
  }

  /**
   * Checks if a node with the specified key exists in the tree.
   * @param key The key of the node to check.
   */
  public contains(key: string): boolean {
    return this.keys().includes(key);
  }

  /**
   * Returns an array of all keys present in the tree.
   */
  public keys(): string[] {
    const keys: string[] = [];

    this.getKeyValuePairs(this.root, '', word => keys.push(word));

    return keys;
  }

  /**
   * Returns all keys present in the tree that begin with the specified prefix.
   * @param prefix The key prefix to search by.
   */
  public keysWithPrefix(prefix: string): string[] {
    const keys: string[] = [];

    const start = this.find(prefix, this.root);
    if (start) this.getKeyValuePairs(start.middle, prefix, word => keys.push(word));

    return keys;
  }

  /**
   * Executes the specified callback at each node in the tree whose key begins with the specified prefix.
   * @param prefix The key prefix to search by.
   * @param callback The callback to execute.
   */
  public searchWithPrefix(prefix: string, callback: (key: string, value: Value) => void): void {
    const start = this.find(prefix, this.root);
    if (start) this.getKeyValuePairs(start.middle, prefix, callback);
  }

  /**
   * Performs a depth-first search of the tree beginning from the root node. Executes the specified callback at each visited node.
   * @param callback The callback to execute.
   */
  public dfs(callback: (key: string, value: Value | null) => void): void {
    this.depthFirstSearch(this.root, callback);
  }

  /**
   * Returns the tree as a string.
   */
  public toString(): string {
    return inspect(this.root, { showHidden: false, depth: null });
  }

  private insert(node: Node<Value> | null, word: string, value: Value): Node<Value> {
    // Respect astral / 32bit / surrogate Unicode characters
    // See https://stackoverflow.com/questions/21397316/split-javascript-string-into-array-of-codepoints-taking-into-account-surrogat
    const key = [...word][0];

    // Create a new node if the input node is null
    if (!node) {
      node = new Node(key);
    }

    // If the root is null then the tree is empty
    if (!this.root) {
      this.root = node;
    }

    // Recurse Left
    if (key < node.key) {
      node.left = this.insert(node.left, word, value);
      node.left.parent = node;
      // Recurse Right
    } else if (key > node.key) {
      node.right = this.insert(node.right, word, value);
      node.right.parent = node;
      // Recurse Middle
    } else if (word.length > 1) {
      node.middle = this.insert(node.middle, word.slice(1), value);
      node.middle.parent = node;
      // End of input
    } else {
      node.value = value;
      this.count += 1;
    }

    return node;
  }

  private find(word: string, node: Node<Value> | null): Node<Value> | null {
    if (!node) return null;

    const key = [...word][0];

    // Recurse Left
    if (key < node.key) {
      return this.find(word, node.left);
      // Recurse Right
    } else if (key > node.key) {
      return this.find(word, node.right);
      // Recurse Middle
    } else if (word.length > 1) {
      return this.find(word.slice(1), node.middle);
      // End of input
    } else {
      return node;
    }
  }

  private delete(node: Node<Value> | null): void {
    if (!node) return;

    if (node.middle) {
      return;
    } else if (!node.middle && !node.left && !node.right && node.value) {
      return;
    } else if (!node.middle && !node.left && !node.right) {
      if (!node.parent) {
        this.root = null;
      } else if (node.parent.left === node) {
        node.parent.left = null;
      } else if (node.parent.middle === node) {
        node.parent.middle = null;
      } else if (node.parent.right === node) {
        node.parent.right = null;
      }

      this.delete(node.parent);
    } else if ((!node.left && !node.middle) || (!node.right && !node.middle)) {
      const child = (node.left || node.right) as Node<Value>;

      if (!node.parent) {
        this.root = child;
      } else if (node.parent.left === node) {
        node.parent.left = child;
      } else if (node.parent.middle === node) {
        node.parent.middle = child;
      } else if (node.parent.right === node) {
        node.parent.right = child;
      }

      child.parent = node.parent;
    } else if (node.right && node.left && !node.middle) {
      let child = node.left;

      while (child.right) {
        child = child.right;
      }

      if (node.left !== child) {
        if (child.parent) {
          child.parent.right = child.left;

          if (child.left) {
            child.left.parent = child.parent;
          }
        }

        child.right = node.right;
        child.left = node.left;
        node.right.parent = child;
        node.left.parent = child;
      } else {
        child.right = node.right;
        child.right.parent = child;
      }

      child.parent = node.parent;

      if (!node.parent) {
        this.root = child;
      } else {
        if (node.parent.left === node) {
          node.parent.left = child;
          child.parent = node.parent;
        } else if (node.parent.middle === node) {
          node.parent.middle = child;
          child.parent = node.parent;
        } else if (node.parent.right === node) {
          node.parent.right = child;
          child.parent = node.parent;
        }
      }
    }
  }

  private getKeyValuePairs(
    node: Node<Value> | null,
    prefix: string,
    callback: (key: string, value: Value) => void,
  ) {
    if (!node) return;

    if (node.value) callback(prefix + node.key, node.value);

    this.getKeyValuePairs(node.left, prefix, callback);
    this.getKeyValuePairs(node.middle, prefix + node.key, callback);
    this.getKeyValuePairs(node.right, prefix, callback);
  }

  private depthFirstSearch(
    node: Node<Value> | null,
    callback: (key: string, value: Value | null) => void,
  ): void {
    if (!node) return;

    callback(node.key, node.value);

    // Left
    if (node.left) {
      this.depthFirstSearch(node.left, callback);
    }

    // Middle
    if (node.middle) {
      this.depthFirstSearch(node.middle, callback);
    }

    // Right
    if (node.right) {
      this.depthFirstSearch(node.right, callback);
    }
  }

  private validateKey(key: string): void {
    if (typeof key !== 'string') {
      throw new TypeError(
        `[Trie] Only string values can be used as keys. Received type ${typeof key}.`,
      );
    }

    if (key.length < 1) {
      throw new Error(`[Trie] Key values must have length greater than or equal to 1.`);
    }
  }
}
