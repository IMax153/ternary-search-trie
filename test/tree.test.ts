import { Trie } from '../src';

describe('Trie', () => {
  describe('Trie.size', () => {
    it('should return the correct size of the tree', () => {
      const trie = new Trie<string>();

      trie.set('foo', 'foo');
      trie.set('ðŸ˜€', 'ðŸ˜€');
      trie.set('!', '!');

      expect(trie.size).toBe(3);
    });

    it('should return zero when the tree is empty', () => {
      const trie = new Trie<string>();

      expect(trie.size).toBe(0);
    });
  });

  describe('Trie.keys', () => {
    it('should return all keys in the tree', () => {
      const trie = new Trie<string>();

      trie.set('foo', 'foo');
      trie.set('ðŸ˜€', 'ðŸ˜€');
      trie.set('!', '!');

      expect(trie.keys()).toContain('f');
      expect(trie.keys()).toContain('o');
      expect(trie.keys()).toContain('ðŸ˜€');
      expect(trie.keys()).toContain('!');
    });

    it('should return an empty array when the tree is empty', () => {
      const trie = new Trie<string>();

      expect(trie.keys()).toEqual([]);
    });
  });

  describe('Trie.set', () => {
    it('should set a key/value pair in the tree', () => {
      const trie = new Trie<string>();

      trie.set('foo', 'foo');
      trie.set('ðŸ˜€', 'ðŸ˜€');
      trie.set('!', '!');

      expect(trie.get('foo')).toBe('foo');
      expect(trie.get('ðŸ˜€')).toBe('ðŸ˜€');
      expect(trie.get('!')).toBe('!');
    });

    it('should not set a key/value pair when provided an invalid key', () => {
      const trie = new Trie<string>();

      // @ts-ignore
      expect(() => trie.set(1, '1')).toThrow(
        '[Trie] Only string values can be used as keys. Received type number.',
      );
      expect(() => trie.set('', '1')).toThrow(
        '[Trie] Key values must have length greater than or equal to 1.',
      );
    });
  });

  describe('Trie.get', () => {
    it('should get a value from the tree based upon the specified key', () => {
      const trie = new Trie<string>();

      trie.set('foo', 'foo');
      trie.set('ðŸ˜€', 'ðŸ˜€');
      trie.set('!', '!');

      expect(trie.get('foo')).toBe('foo');
      expect(trie.get('ðŸ˜€')).toBe('ðŸ˜€');
      expect(trie.get('!')).toBe('!');
    });

    it('should not get a value when provided an invalid key', () => {
      const trie = new Trie<string>();

      // @ts-ignore
      expect(() => trie.get(1)).toThrow(
        '[Trie] Only string values can be used as keys. Received type number.',
      );
      expect(() => trie.get('')).toThrow(
        '[Trie] Key values must have length greater than or equal to 1.',
      );
    });
  });

  describe('Trie.del', () => {
    it('should delete a node from the tree', () => {
      const trie = new Trie<string>();

      trie.set('foo', 'foo');
      trie.set('fooooo', 'fooooo');
      trie.set('bar', 'bar');
      trie.set('baz', 'baz');
      trie.set('ðŸ˜€', 'ðŸ˜€');
      trie.set('!', '!');

      expect(trie.get('foo')).toBe('foo');
      expect(trie.get('fooooo')).toBe('fooooo');
      expect(trie.get('bar')).toBe('bar');
      expect(trie.get('baz')).toBe('baz');

      trie.del('foo');
      trie.del('baz');

      expect(trie.get('foo')).toBe(null);
      expect(trie.get('baz')).toBe(null);
      expect(trie.get('bar')).toBe('bar');
      expect(trie.get('fooooo')).toBe('fooooo');
    });
  });

  describe('Trie.traverse', () => {
    it('should traverse each node in the tree', () => {
      const trie = new Trie<string>();
      const keys = ['foo', 'ðŸ˜€', '!', 'fooo', 'bar'];

      keys.forEach(key => trie.set(key, key));

      trie.traverse(node => {
        expect(keys.some(key => key.includes(node.key))).toBeTruthy();

        if (node.value) {
          expect(keys).toContain(node.value);
        }
      });
    });
  });

  describe('Trie.searchByPrefix', () => {
    it('shoud return a list of words found using the specified prefix', () => {
      const trie = new Trie<string>();

      trie.set('foo', 'foo');
      trie.set('fooooo', 'fooooo');
      trie.set('fore', 'fore');
      trie.set('fobe', 'fobe');
      trie.set('bar', 'bar');
      trie.set('baz', 'baz');
      trie.set('ðŸ˜€', 'ðŸ˜€');
      trie.set('!', '!');

      const expected = ['foo', 'fore', 'fobe', 'fooooo'];

      trie.searchByPrefix('fo', word => {
        expect(expected).toContain(word);
      });

      trie.searchByPrefix('ðŸ˜€', word => {
        expect(word).toBe('ðŸ˜€');
      });
    });
  });

  describe('Trie.toString', () => {
    it('should correctly print the tree', () => {
      const trie = new Trie<string>();

      trie.set('foo', 'foo');
      trie.set('fooooo', 'fooooo');
      trie.set('bar', 'bar');
      trie.set('baz', 'baz');
      trie.set('ðŸ˜€', 'ðŸ˜€');
      trie.set('!', '!');

      expect(trie.toString()).toMatchSnapshot();
    });
  });
});
