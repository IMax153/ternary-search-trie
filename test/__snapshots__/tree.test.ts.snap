// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trie Trie.toString should correctly print the tree 1`] = `
"├─ key
├─ value
├─ parent
├─ left
│  ├─ key
│  ├─ value
│  ├─ parent (circular ref.)
│  ├─ left
│  │  ├─ key
│  │  ├─ value
│  │  ├─ parent (circular ref.)
│  │  ├─ left
│  │  ├─ middle
│  │  └─ right
│  ├─ middle
│  │  ├─ key
│  │  ├─ value
│  │  ├─ parent (circular ref.)
│  │  ├─ left
│  │  ├─ middle
│  │  │  ├─ key
│  │  │  ├─ value
│  │  │  ├─ parent (circular ref.)
│  │  │  ├─ left
│  │  │  ├─ middle
│  │  │  └─ right
│  │  │     ├─ key
│  │  │     ├─ value
│  │  │     ├─ parent (circular ref.)
│  │  │     ├─ left
│  │  │     ├─ middle
│  │  │     └─ right
│  │  └─ right
│  └─ right
├─ middle
│  ├─ key
│  ├─ value
│  ├─ parent (circular ref.)
│  ├─ left
│  ├─ middle
│  │  ├─ key
│  │  ├─ value
│  │  ├─ parent (circular ref.)
│  │  ├─ left
│  │  ├─ middle
│  │  │  ├─ key
│  │  │  ├─ value
│  │  │  ├─ parent (circular ref.)
│  │  │  ├─ left
│  │  │  ├─ middle
│  │  │  │  ├─ key
│  │  │  │  ├─ value
│  │  │  │  ├─ parent (circular ref.)
│  │  │  │  ├─ left
│  │  │  │  ├─ middle
│  │  │  │  │  ├─ key
│  │  │  │  │  ├─ value
│  │  │  │  │  ├─ parent (circular ref.)
│  │  │  │  │  ├─ left
│  │  │  │  │  ├─ middle
│  │  │  │  │  └─ right
│  │  │  │  └─ right
│  │  │  └─ right
│  │  └─ right
│  └─ right
└─ right
   ├─ key
   ├─ value
   ├─ parent (circular ref.)
   ├─ left
   ├─ middle
   │  ├─ key
   │  ├─ value
   │  ├─ parent (circular ref.)
   │  ├─ left
   │  ├─ middle
   │  └─ right
   └─ right
"
`;
